pkg/deploy/elbv2/target_group_synthesizer.go

```diff
@@ -0,0 +1,169 @@
+func (s *targetGroupSynthesizer) findSDKTargetGroups(ctx context.Context) ([]TargetGroupWithTags, error) {
+	stackTags := s.taggingProvider.StackTags(s.stack)
+	return s.taggingManager.ListTargetGroups(ctx, tagging.TagsAsMultiValueTagFilter(stackTags))
+}
+
+type resAndSDKTargetGroupPair struct {
+	resTG *elbv2model.TargetGroup
+	sdkTG TargetGroupWithTags
+}
+
+func matchResAndSDKTargetGroups(resTGs []*elbv2model.TargetGroup, sdkTGs []TargetGroupWithTags,
+	resourceIDTagKey string) ([]resAndSDKTargetGroupPair, []*elbv2model.TargetGroup, []TargetGroupWithTags, error) {
+	var matchedResAndSDKTGs []resAndSDKTargetGroupPair
+	var unmatchedResTGs []*elbv2model.TargetGroup
+	var unmatchedSDKTGs []TargetGroupWithTags
+
+	resTGsByID := mapResTargetGroupByResourceID(resTGs)
+	sdkTGsByID, err := mapSDKTargetGroupByResourceID(sdkTGs, resourceIDTagKey)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	resTGIDs := sets.StringKeySet(resTGsByID)
+	sdkTGIDs := sets.StringKeySet(sdkTGsByID)
+	for _, resID := range resTGIDs.Intersection(sdkTGIDs).List() {
+		resTG := resTGsByID[resID]
+		sdkTGs := sdkTGsByID[resID]
+		foundMatch := false
+		for _, sdkTG := range sdkTGs {
+			if isSDKTargetGroupRequiresReplacement(sdkTG, resTG) {
+				unmatchedSDKTGs = append(unmatchedSDKTGs, sdkTG)
+				continue
+			}
+			matchedResAndSDKTGs = append(matchedResAndSDKTGs, resAndSDKTargetGroupPair{
+				resTG: resTG,
+				sdkTG: sdkTG,
+			})
+			foundMatch = true
+		}
+		if !foundMatch {
+			unmatchedResTGs = append(unmatchedResTGs, resTG)
+		}
+	}
+	for _, resID := range resTGIDs.Difference(sdkTGIDs).List() {
+		unmatchedResTGs = append(unmatchedResTGs, resTGsByID[resID])
+	}
+	for _, resID := range sdkTGIDs.Difference(resTGIDs).List() {
+		unmatchedSDKTGs = append(unmatchedSDKTGs, sdkTGsByID[resID]...)
+	}
+
+	return matchedResAndSDKTGs, unmatchedResTGs, unmatchedSDKTGs, nil
+}
+
+func mapResTargetGroupByResourceID(resTGs []*elbv2model.TargetGroup) map[string]*elbv2model.TargetGroup {
+	resTGsByID := make(map[string]*elbv2model.TargetGroup, len(resTGs))
+	for _, resTG := range resTGs {
+		resTGsByID[resTG.ID()] = resTG
+	}
+	return resTGsByID
+}
+
+func mapSDKTargetGroupByResourceID(sdkTGs []TargetGroupWithTags, resourceIDTagKey string) (map[string][]TargetGroupWithTags, error) {
+	sdkTGsByID := make(map[string][]TargetGroupWithTags, len(sdkTGs))
+	for _, sdkTG := range sdkTGs {
+		resourceID, ok := sdkTG.Tags[resourceIDTagKey]
+		if !ok {
+			return nil, errors.Errorf("unexpected targetGroup with no resourceID: %v", awssdk.StringValue(sdkTG.TargetGroup.TargetGroupArn))
+		}
+		sdkTGsByID[resourceID] = append(sdkTGsByID[resourceID], sdkTG)
+	}
+	return sdkTGsByID, nil
+}
+
+// isSDKTargetGroupRequiresReplacement checks whether a sdk TargetGroup requires replacement to fulfill a TargetGroup resource.
+func isSDKTargetGroupRequiresReplacement(sdkTG TargetGroupWithTags, resTG *elbv2model.TargetGroup) bool {
```

Comment:
Currently, it is not possible to edit health check protocol and interval if the target group protocol is TCP, in case of NLB. How are those tgt groups to be handled?

Reply:
nice catch 👍 
added check for nlb's healthCheck immutability.

When you build the name of TargetGroup, you need to include these immutable fields as part the targetGroup. then it will be handled automatically.  (The TargetGroup resource ID can still be same).